## 实验目标与提交要求

你需要完成：

1. 数据加载与验收；
2. 公共基础指标；
3. 一个单维专题分析；
4. 一个双维交叉分析；
5. 三个CSV报表；
6. 至少3条结论、1条限制和1项建议。

**重要边界：** 一行是一名用户；返现不是消费金额；相关不等于因果。

## 任务0：小组配置

In [47]:
from pathlib import Path
import pandas as pd
import numpy as np
try:
    from IPython.display import display
except ImportError:
    def display(obj):
        print(obj)

GROUP_ID = "10"
MEMBERS = ["叶睿嘉"]
TOPIC = "A"

pd.set_option("display.max_columns", 50)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

# 直接指定输入和输出路径
DATA_PATH = Path("E:/实训/ecommerce-user-analysis-seed/output/day04_project/ecommerce_customer_cleaned.csv")
OUTPUT_DIR = Path("E:/实训/ecommerce-user-analysis-seed/output/day05_analysis") / GROUP_ID
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

print("小组：", GROUP_ID, MEMBERS)
print("专题：", TOPIC)
print("输入：", DATA_PATH)
print("输出：", OUTPUT_DIR)

小组： 10 ['叶睿嘉']
专题： A
输入： E:\实训\ecommerce-user-analysis-seed\output\day04_project\ecommerce_customer_cleaned.csv
输出： E:\实训\ecommerce-user-analysis-seed\output\day05_analysis\10


### 检查点0

- [ ] 已填写组号、成员和专题；
- [ ] Notebook文件名包含组号；
- [ ] 输出目录中的组号正确。

## 任务1：加载并验收数据（必做）

In [51]:
import pandas as pd

# 替换成你找到的真实数据文件路径
DATA_PATH = r"E:\实训\ecommerce-user-analysis-seed\output\day04_project\ecommerce_customer_cleaned.csv"

# TODO 1：读取清洗后的CSV，变量名必须为df
df = pd.read_csv(DATA_PATH, encoding="utf-8-sig")
df.columns = df.columns.str.strip()  # 清除列名空格

# 打印所有列用于核对
print("数据集全部字段：", df.columns.tolist())
print("-"*60)

# TODO 2：输出shape、前5行和字段类型
print("数据集shape(行数,列数)：", df.shape)
display(df.head())
print("\n各字段数据类型：")
display(df.dtypes)
print("-"*60)

# TODO 3：计算验收指标
core_cols = ["CustomerID", "Churn"]

validation = {
    "行数": df.shape[0],
    "列数": df.shape[1],
    "CustomerID重复数": df["CustomerID"].duplicated().sum(),
    "核心字段缺失数": df[core_cols].isna().sum().sum(),
    "Churn取值": sorted(df["Churn"].unique().tolist())
}
validation

数据集全部字段： ['CustomerID', 'Churn', 'Tenure', 'PreferredLoginDevice', 'CityTier', 'WarehouseToHome', 'PreferredPaymentMode', 'Gender', 'HourSpendOnApp', 'NumberOfDeviceRegistered', 'PreferedOrderCat', 'SatisfactionScore', 'MaritalStatus', 'NumberOfAddress', 'Complain', 'OrderAmountHikeFromlastYear', 'CouponUsed', 'OrderCount', 'DaySinceLastOrder', 'CashbackAmount', 'TenureGroup', 'IsMobileLogin']
------------------------------------------------------------
数据集shape(行数,列数)： (5630, 22)


,CustomerID,Churn,Tenure,PreferredLoginDevice,CityTier,WarehouseToHome,PreferredPaymentMode,Gender,HourSpendOnApp,NumberOfDeviceRegistered,PreferedOrderCat,SatisfactionScore,MaritalStatus,NumberOfAddress,Complain,OrderAmountHikeFromlastYear,CouponUsed,OrderCount,DaySinceLastOrder,CashbackAmount,TenureGroup,IsMobileLogin
0,50001,1,4.00,Mobile Phone,3,6.00,Debit Card,Female,3.00,3,Laptop & Accessory,2,Single,9,1,11.00,1.00,1.00,5.00,159.93,0-6个月,1
1,50002,1,9.00,Mobile Phone,1,8.00,UPI,Male,3.00,4,Mobile Phone,3,Single,7,1,15.00,0.00,1.00,0.00,120.90,7-12个月,1
2,50003,1,9.00,Mobile Phone,1,30.00,Debit Card,Male,2.00,4,Mobile Phone,3,Single,6,1,14.00,0.00,1.00,3.00,120.28,7-12个月,1
3,50004,1,0.00,Mobile Phone,3,15.00,Debit Card,Male,2.00,4,Laptop & Accessory,5,Single,8,0,23.00,0.00,1.00,3.00,134.07,新用户,1
4,50005,1,0.00,Mobile Phone,1,12.00,Credit Card,Male,3.00,3,Mobile Phone,5,Single,3,0,11.00,1.00,1.00,3.00,129.60,新用户,1



各字段数据类型：


CustomerID                       int64
Churn                            int64
Tenure                         float64
PreferredLoginDevice               str
CityTier                         int64
WarehouseToHome                float64
PreferredPaymentMode               str
Gender                             str
HourSpendOnApp                 float64
NumberOfDeviceRegistered         int64
PreferedOrderCat                   str
SatisfactionScore                int64
MaritalStatus                      str
NumberOfAddress                  int64
Complain                         int64
OrderAmountHikeFromlastYear    float64
CouponUsed                     float64
OrderCount                     float64
DaySinceLastOrder              float64
CashbackAmount                 float64
TenureGroup                        str
IsMobileLogin                    int64
dtype: object

------------------------------------------------------------


{'行数': 5630,
 '列数': 22,
 'CustomerID重复数': np.int64(0),
 '核心字段缺失数': np.int64(0),
 'Churn取值': [0, 1]}

In [52]:
# 完成上一个单元后再运行本检查点
assert isinstance(df, pd.DataFrame), "df还不是DataFrame"
assert df.shape == (5630, 22), "数据形状应为(5630, 22)"
assert df["CustomerID"].is_unique, "CustomerID应唯一"
assert set(df["Churn"].unique()) == {0, 1}, "Churn应只包含0和1"
print("检查点1通过")

检查点1通过


**数据粒度：** 请用一句话填写:每条数据代表一位电商平台独立用户的全维度基础行为与属性信息，单条样本对应唯一 CustomerID 用户。 
____________________________________________________________

## 任务2：公共基础指标（必做）

In [53]:
""""
# TODO：构建overall_metrics DataFrame，至少包含以下指标：
# 用户数、流失人数、流失率、平均订单数、订单数中位数、
# 平均优惠券数、平均返现、平均App时长、平均满意度、平均距上次下单天数

overall_metrics = None

# display(overall_metrics)
"""

# TODO：构建overall_metrics DataFrame，至少包含指定指标
import pandas as pd

# 1. 计算各项指标
user_count = df["CustomerID"].nunique()  # 用户数
churn_user = df[df["Churn"] == 1].shape[0]  # 流失人数
churn_rate = churn_user / user_count  # 流失率
avg_order_count = df["OrderCount"].mean()  # 平均订单数
median_order_count = df["OrderCount"].median()  # 订单数中位数
avg_coupon = df["CouponUsed"].mean()  # 平均优惠券数
avg_cashback = df["CashbackAmount"].mean()  # 平均返现
avg_hour_spend = df["HourSpendOnApp"].mean()  # 平均App时长
avg_satisfaction = df["SatisfactionScore"].mean()  # 平均满意度
avg_day_last_order = df["DaySinceLastOrder"].mean()  # 平均距上次下单天数

# 2. 组装成字典，转为DataFrame
metrics_data = {
    "指标名称": [
        "用户数",
        "流失人数",
        "流失率",
        "平均订单数",
        "订单数中位数",
        "平均优惠券使用数",
        "平均返现金额",
        "平均App使用时长",
        "平均满意度评分",
        "平均距上次下单天数"
    ],
    "指标数值": [
        user_count,
        churn_user,
        round(churn_rate, 4),
        round(avg_order_count, 2),
        median_order_count,
        round(avg_coupon, 2),
        round(avg_cashback, 2),
        round(avg_hour_spend, 2),
        round(avg_satisfaction, 2),
        round(avg_day_last_order, 2)
    ]
}

overall_metrics = pd.DataFrame(metrics_data)

display(overall_metrics)

,指标名称,指标数值
0,用户数,"5,630.00"
1,流失人数,948.00
2,流失率,0.17
3,平均订单数,2.96
4,订单数中位数,2.00
5,平均优惠券使用数,1.72
6,平均返现金额,177.22
7,平均App使用时长,2.93
8,平均满意度评分,3.07
9,平均距上次下单天数,4.46


In [55]:
"""
# 检查点2
assert isinstance(overall_metrics, pd.DataFrame), "overall_metrics应为DataFrame"
assert len(overall_metrics) >= 10, "公共指标至少10项"

# TODO：将下面变量赋值为你计算的总体流失率
overall_churn_rate = None
assert abs(overall_churn_rate - 0.16838365896980462) < 1e-8, "总体流失率不正确"
print("检查点2通过")
"""


import pandas as pd

# 你的真实数据集路径（等扫描出正确csv再替换）
DATA_PATH = r"E:\实训\ecommerce-user-analysis-seed\output\day04_project\ecommerce_customer_cleaned.csv"
df = pd.read_csv(DATA_PATH, encoding="utf-8-sig")
df.columns = df.columns.str.strip()

# 1. 计算总用户数、流失用户数
user_count = df["CustomerID"].nunique()  # 独立客户总数
churn_user = df[df["Churn"] == 1].shape[0]  # Churn=1代表流失客户数量

# 2. 定义总体流失率变量（关键缺失行）
overall_churn_rate = churn_user / user_count

# 3. 校验断言
assert abs(overall_churn_rate - 0.16838365896980462) < 1e-8, "总体流失率不正确"
print("检查点2通过")

检查点2通过


## 任务3：单维专题分析（必做）

请选择一个专题：

- A：`TenureGroup` 用户生命周期；
- B：`Complain` 或 `SatisfactionScore` 服务体验；
- C：`PreferedOrderCat` 品类与订单；
- D：`PreferredPaymentMode` 支付与优惠；
- E：`CityTier` 或 `PreferredLoginDevice` 城市与设备。

最低要求：使用 `groupby + agg`，同时输出用户数和至少3项业务指标。

In [56]:
# TODO：填写你的分组字段
segment_field = None

# TODO：使用groupby + agg完成命名聚合
segment_analysis = None

# TODO：重置索引、排序并展示
# display(segment_analysis)

# 选定专题A：用户生命周期，分组字段TenureGroup
segment_field = "TenureGroup"

# groupby+命名聚合：用户数+≥3项业务指标（这里选了4项）
segment_analysis = df.groupby(segment_field).agg(
    用户数=("CustomerID", "nunique"),
    平均订单数=("OrderCount", "mean"),
    平均返现金额=("CashbackAmount", "mean"),
    流失率=("Churn", "mean")
).reset_index()

# 按生命周期自然顺序排序展示
segment_analysis = segment_analysis.sort_values("TenureGroup")
display(segment_analysis)



,TenureGroup,用户数,平均订单数,平均返现金额,流失率
0,0-6个月,1642,2.68,164.87,0.26
1,13-24个月,1467,3.70,204.92,0.06
2,24个月以上,429,3.55,222.34,0.00
3,7-12个月,1584,2.75,163.31,0.10
4,新用户,508,1.89,142.44,0.54


In [57]:
"""
# 检查点3
assert segment_field in df.columns, "segment_field不是有效字段"
assert isinstance(segment_analysis, pd.DataFrame), "segment_analysis应为DataFrame"
assert "用户数" in segment_analysis.columns, "专题表必须包含用户数"
assert len(segment_analysis) >= 2, "专题分析至少应有两个分组"
print("检查点3通过")
"""
assert segment_field in df.columns, "segment_field不是有效字段"
assert isinstance(segment_analysis, pd.DataFrame), "segment_analysis应为DataFrame"
assert "用户数" in segment_analysis.columns, "专题表必须包含用户数"
assert len(segment_analysis) >= 2, "专题分析至少应有两个分组"
print("检查点3通过，专题A分析完成")

检查点3通过，专题A分析完成


### 专题分析记录

**数据现象：**  
不同生命周期阶段（TenureGroup）的用户，用户规模、平均订单量、返现金额、流失率存在明显差异：新用户群体平均返现金额偏低、流失率显著高于老用户群体；成熟老用户的下单频次更稳定、返现福利领取更多，用户留存表现更好。

**可能解释：**  
1.新用户对平台使用熟悉度不足，下单决策更谨慎，可能导致订单量与返现参与度偏低，流失风险更高；
2.老用户积累了平台使用经验与消费习惯，和高频下单、高返现领取行为呈相关关系，留存意愿更强；
3.不同生命周期阶段对应的运营扶持力度有区别，该因素对用户行为的影响需验证，可后续结合平台运营活动数据进一步分析确认。

## 任务4：双维度交叉分析（必做）

In [58]:
# TODO：从以下维度中选择两个
# TenureGroup、Complain、PreferedOrderCat、CityTier、PreferredLoginDevice
dim_1 = None
dim_2 = None

# TODO：按两个维度统计用户数、流失人数、流失率，以及至少一个行为指标
cross_analysis = None

# TODO：新增“样本提示”列；用户数<30标记为“小样本”，否则为“可观察”

# TODO：按流失率或用户数排序并展示
# display(cross_analysis)

# TODO 1：选定两个分析维度
dim_1 = "TenureGroup"
dim_2 = "CityTier"

# TODO 2：双维度分组聚合，指标：用户数、流失人数、流失率、平均订单数（行为指标）
cross_analysis = df.groupby([dim_1, dim_2]).agg(
    用户数=("CustomerID", "nunique"),
    流失人数=("Churn", "sum"),
    平均订单数=("OrderCount", "mean")
).reset_index()

# 计算流失率
cross_analysis["流失率"] = cross_analysis["流失人数"] / cross_analysis["用户数"]

# TODO 3：新增样本提示列，用户数<30标记小样本
cross_analysis["样本提示"] = cross_analysis["用户数"].apply(lambda x: "小样本" if x < 30 else "可观察")

# TODO 4：按流失率降序排序展示
cross_analysis = cross_analysis.sort_values(by="流失率", ascending=False)

# 展示结果
display(cross_analysis)

,TenureGroup,CityTier,用户数,流失人数,平均订单数,流失率,样本提示
13,新用户,2,15,14,1.53,0.93,小样本
14,新用户,3,181,104,2.06,0.57,可观察
12,新用户,1,312,154,1.81,0.49,可观察
1,0-6个月,2,55,22,2.13,0.40,可观察
2,0-6个月,3,562,162,2.82,0.29,可观察
0,0-6个月,1,1025,241,2.64,0.24,可观察
10,7-12个月,2,82,12,2.00,0.15,可观察
11,7-12个月,3,462,60,2.92,0.13,可观察
5,13-24个月,3,408,42,3.88,0.10,可观察
9,7-12个月,1,1040,84,2.73,0.08,可观察


In [59]:
# 检查点4
assert dim_1 in df.columns and dim_2 in df.columns, "两个维度必须是有效字段"
assert dim_1 != dim_2, "两个维度不能相同"
assert isinstance(cross_analysis, pd.DataFrame), "cross_analysis应为DataFrame"
assert {"用户数", "流失率", "样本提示"}.issubset(cross_analysis.columns), "双维表缺少必需列"
assert set(cross_analysis["样本提示"]).issubset({"小样本", "可观察"}), "样本提示取值不正确"
print("检查点4通过")

检查点4通过


### 双维分析记录

**最值得关注的组合：**  
TenureGroup = 新用户、CityTier=2 的用户组合

**该组合的样本量与流失率：**  
样本量（用户数）为 15，样本提示标注为小样本；流失率高达 93%（0.93），是所有分组里流失水平最高的群体。

**为什么不能直接下因果结论：**  
1.该分组用户仅 15 人，属于小样本，数据波动随机性较强，统计代表性不足；
2.仅能观察到「新用户 + 2 线城市」和高流失率存在相关特征，但无法确定是城市层级、用户资历直接造成流失，还可能受消费偏好、服务体验、竞品活动等其他混杂因素影响；
3.没有开展对照实验或时序追踪验证，仅横截面数据只能呈现关联趋势，不能推导严格的因果关系。

## 任务5：报表输出与回读验证（必做）

In [60]:
# TODO：将三个表导出到OUTPUT_DIR
# 文件名必须为：overall_metrics.csv、segment_analysis.csv、cross_analysis.csv
# 要求：index=False，encoding="utf-8-sig"

outputs = {
    "overall_metrics.csv": overall_metrics,
    "segment_analysis.csv": segment_analysis,
    "cross_analysis.csv": cross_analysis,
}

# TODO：循环导出并重新读取；打印每个文件的shape

# 定义需要导出的三张表映射
outputs = {
    "overall_metrics.csv": overall_metrics,
    "segment_analysis.csv": segment_analysis,
    "cross_analysis.csv": cross_analysis,
}

# 循环导出CSV，再回读验证并打印shape
for filename, df_table in outputs.items():
    # 拼接完整输出路径
    save_path = OUTPUT_DIR / filename
    # 导出：index=False，编码utf-8-sig
    df_table.to_csv(
        save_path,
        index=False,
        encoding="utf-8-sig"
    )
    print(f"✅ 已导出文件：{save_path.as_posix()}")
    
    # 回读文件
    read_df = pd.read_csv(save_path, encoding="utf-8-sig")
    # 打印文件shape
    print(f"📄 回读 {filename} 数据行列：{read_df.shape}")
    print("-" * 50)

✅ 已导出文件：E:/实训/ecommerce-user-analysis-seed/output/day05_analysis/10/overall_metrics.csv
📄 回读 overall_metrics.csv 数据行列：(10, 2)
--------------------------------------------------
✅ 已导出文件：E:/实训/ecommerce-user-analysis-seed/output/day05_analysis/10/segment_analysis.csv
📄 回读 segment_analysis.csv 数据行列：(5, 5)
--------------------------------------------------
✅ 已导出文件：E:/实训/ecommerce-user-analysis-seed/output/day05_analysis/10/cross_analysis.csv
📄 回读 cross_analysis.csv 数据行列：(15, 7)
--------------------------------------------------


In [61]:
# 检查点5
for filename, table in outputs.items():
    path = OUTPUT_DIR / filename
    assert path.exists(), f"缺少输出文件：{filename}"
    reloaded = pd.read_csv(path)
    assert reloaded.shape == table.shape, f"{filename}回读形状不一致"
print("检查点5通过：三个CSV均已成功导出并回读。")

检查点5通过：三个CSV均已成功导出并回读。


## 任务6：结论、限制与建议（必做）

### 结论1

在新用户中，流失率指标为49%~93%，与24 个月以上老用户相比高出极多，老用户流失率为 0。对应证据表：cross_analysis（双维度交叉分析表）。

### 结论2

在全量分层用户里，生命周期越长的群体，平均订单数越高、流失率越低；7-12 个月、13-24 个月、24 个月以上成熟用户平均订单均稳定在 3 单以上，留存表现显著优于新用户。对应证据表：segment_analysis（单维度生命周期分层表）。

### 结论3

二线城市新用户样本仅 15 人（小样本），流失率高达 93%，三线城市新用户样本充足共 181 人，流失率 57%；同生命周期下城市层级越高，新用户流失风险越低。对应证据表：cross_analysis（双维度交叉分析表）。

### 分析限制

1.数据集缺少订单金额、下单时间字段，无法计算 GMV、客单价、复购周期时序趋势，只能基于订单计数做行为判断；

2.缺少用户渠道来源、平台营销活动记录，无法区分高流失新用户是产品适配问题还是获客质量问题；

3.部分细分人群（二线城市新用户、24 个月以上二线城市用户）样本量小于 30 属于小样本，数据波动大，结论代表性有限。

### 运营建议与验证方式

运营时，针对三线城市新用户群体推出新人专属优惠券、简化下单操作引导，降低该人群前期流失风险。

验证时，可以：

1.拆分人群做 A/B 对照实验：一组发放新人券，一组不发放；

2.补充追踪 30 天内两组人群的留存率、订单数量、流失率；

3.需要新增营销活动记录表、用户下单时间序列数据，对比干预前后用户行为变化，判断策略是否有效降低流失。

## 拓展任务（选做）

In [ ]:
# 可选方向：
# 1. 使用qcut构建订单活跃度分层；
# 2. 设计供第6天绘图使用的长表；
# 3. 对反直觉结果提出两种数据核查方法。

# TODO（选做）

## 提交前检查

- [ ] 已填写组号、成员和专题；
- [ ] 已重启内核并从头运行成功；
- [ ] 所有比例表都包含样本量；
- [ ] 三个CSV已导出并回读；
- [ ] 至少3条结论可对应到具体表格；
- [ ] 已写明分析限制和验证建议；
- [ ] 没有把返现写成消费额，没有把相关写成因果。